internal/tuning.go

```diff
@@ -0,0 +1,266 @@
+// The MIT License
+//
+// Copyright (c) 2022 Temporal Technologies Inc.  All rights reserved.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package internal
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"sync/atomic"
+
+	"golang.org/x/sync/semaphore"
+
+	"go.temporal.io/sdk/internal/common/metrics"
+)
+
+// WorkerTuner allows for the dynamic customization of some aspects of worker behavior.
+//
+// WARNING: Custom implementations of SlotSupplier are currently experimental.
+type WorkerTuner interface {
+	GetWorkflowTaskSlotSupplier() SlotSupplier
+	GetActivityTaskSlotSupplier() SlotSupplier
+	GetLocalActivitySlotSupplier() SlotSupplier
+}
+
+// SlotPermit is a permit to use a slot.
+//
+// WARNING: Custom implementations of SlotSupplier are currently experimental.
+type SlotPermit struct {
+	//lint:ignore U1000 pointless to guarantee uniqueness for now
+	int
+}
+
+// SlotReserveContext contains information that SlotSupplier instances can use during
+// reservation calls.
+type SlotReserveContext interface {
+	TaskQueue() string
+	NumIssuedSlots() int
+}
+
+// SlotSupplier controls how slots are handed out for workflow and activity tasks as well as
+// local activities when used in conjunction with a WorkerTuner.
+//
+// Currently, you cannot implement your own slot supplier. You can use the provided
+// FixedSizeSlotSupplier and ResourceBasedSlotSupplier slot suppliers.
+//
+// WARNING: Custom implementations of SlotSupplier are currently experimental.
+type SlotSupplier interface {
+	// ReserveSlot is called before polling for new tasks. The implementation should block until
+	// a slot is available, then return a permit to use that slot. Implementations must be
+	// thread-safe.
+	//
+	// Any returned error besides context.Canceled will be logged and the function will be retried.
+	ReserveSlot(ctx context.Context, reserveCtx SlotReserveContext) (*SlotPermit, error)
```

Comment:
Yeah, my concern was implying that they could be cancelled. I figure if they want to pass information, they can just do it via whatever struct is implementing this interface.

Reply:
The issue is contextual state per reservation. I don't think people see Go context as meaning async. In fact, since reserve context is rarely used, would totally support just a Go context and a method to retrieve the reserve context from it.

Also, arguably `SlotPermit` should be an (empty) interface so that people can pass back their own objects that are sent back on mark used and release. People might want to attach a state to a slot (even if you don't do that, should at least provide the slot permit to mark used and release so they can keep a map).

Reply:
Retrieving it from the Go context just feels icky to me, how it introduces a number of possible error paths.

I'd be fine with sticking the Go context inside this context though, and since all the other methods will get their own context then they'll have that.

Attaching state to the permit was done in Java by just having a field inside `SlotPermit`, which I also prefer since it allows us to be clear about what the type of a permit is instead of it just being `any`. I just didn't add it since it's unclear if anyone even wants that at the moment, but, it's trivial to add to this too.

Reply:
Concur, don't need to retrieve from Go context (can keep separate or, how I like, embed the Go context in the slot reserve context). My only remaining concern is using Go contexts in non-cancellable/non-long-lived situations like "try". I think we should as Go context is for more than just long-running functions.